src/blockdev.rs

```diff
@@ -636,6 +636,15 @@ impl SavedPartitions {
 
         // write
         gpt.write_into(disk).chain_err(|| "writing updated GPT")?;
+
+        // If the MBR signature is missing, write a protective MBR, otherwise
+        // leave alone.  The kernel will ignore a GPT without some sort of
+        // MBR.
+        if !disk_has_mbr(disk)? {
```

Comment:
In the failure case, we probably can't restore the full table because we've overwritten some of the disk.  In particular, we don't want to restore any boot code in the MBR.  So instead we create a fresh table (including protective MBR and a new disk GUID) and preserve only those partitions we were asked to preserve.

Reply:
Ack, makes sense.